core/shared/src/main/scala/scalaz/zio/internal/Executor.scala

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2018 - 2019 John A. De Goes. All rights reserved.
+package scalaz.zio.internal
+
+import java.util.concurrent._
+import scala.concurrent.ExecutionContext
+
+/**
+ * An executor is responsible for executing actions. Each action is guaranteed
+ * to begin execution on a fresh stack frame.
+ */
+trait Executor {
+
+  /**
+   * The concurrency level of the executor.
+   */
+  def concurrency: Int
+
+  /**
+   * The capacity of the executor.
+   */
+  def capacity: Int
+
+  /**
+   * The number of tasks remaining to be executed.
+   */
+  def size: Int
+
+  /**
+   * The number of tasks that have been enqueued, over all time.
+   */
+  def enqueuedCount: Long
+
+  /**
+   * The number of tasks that have been dequeued, over all time.
+   */
+  def dequeuedCount: Long
+
+  /**
+   * Submits a task for execution.
+   */
+  def submit(runnable: Runnable): Boolean
+
+  /**
+   * Initiates shutdown of the executor.
+   */
+  def shutdown(): Unit
+
+  /**
+   * Views this `Executor` as a Scala `ExecutionContext`.
+   */
+  lazy val asExecutionContext: ExecutionContext =
+    new ExecutionContext {
+      override def execute(r: Runnable): Unit =
+        if (!submit(r)) throw new RejectedExecutionException("Rejected: " + r.toString)
+
+      override def reportFailure(cause: Throwable): Unit =
+        cause.printStackTrace
```

Comment:
to logger? or reporter?

Reply:
No one should ever call this... it's certainly not called internally. `ExecutionContext` just forces us to implement it.